File: tests/golden/orm/jpa-java/postgresql/house-defaults.java
/*
 * grabado が生成した Jakarta Persistence の entity（Java）。
 *
 * **1 クラス 1 ファイルに分けてから使う。** Java は 1 つのコンパイル単位に public な
 * クラスを 1 つしか置けない。区切りの行が次のファイルの始まりで、そこに書いてある名前が
 * ファイル名になる。import はファイルごとに付けてあるので、切ればそのまま通る。
 *
 * package 宣言は出さない（置き場所は生成物を受け取る側が決める）。
 * **コンストラクタも出していない** —— JPA は引数の無いコンストラクタを要求するが、
 * 明示のコンストラクタが 1 つも無ければ Java がそれを作る。**足すときは引数無しも残すこと。**
 *
 * 型はすべてボクシング型（int ではなく Integer）。primitive は null を表せないので、
 * outer join や部分ロードで壊れる。NOT NULL は @Column(nullable = false) が表す。
 */

/* ==== User.java ==== */

import jakarta.persistence.Column;
import jakarta.persistence.Entity;
import jakarta.persistence.Id;
import jakarta.persistence.Table;
import jakarta.persistence.UniqueConstraint;
import java.time.OffsetDateTime;
import java.util.UUID;

/** ユーザー */
@Entity
@Table(name = "users", uniqueConstraints = {@UniqueConstraint(name = "users_email_key", columnNames = {"email"})})
public class User {
    @Id
    @Column(name = "id")
    private UUID id;

    /** ログイン用メールアドレス */
    @Column(name = "email", nullable = false)
    private String email;

    @Column(name = "display_name", nullable = false)
    private String displayName;

    @Column(name = "is_active", nullable = false)
    private Boolean isActive;

    /** UI 設定などの任意項目 */
    @Column(name = "preferences", nullable = false)
    /* json: JPA の標準に対応する型が無いので String で出す（JSONB） */
    private String preferences;

    @Column(name = "created_at", nullable = false)
    private OffsetDateTime createdAt;

    @Column(name = "updated_at", nullable = false)
    private OffsetDateTime updatedAt;

    public UUID getId() {
        return id;
    }

    public void setId(UUID id) {
        this.id = id;
    }

    public String getEmail() {
        return email;
    }

    public void setEmail(String email) {
        this.email = email;
    }

    public String getDisplayName() {
        return displayName;
    }

    public void setDisplayName(String displayName) {
        this.displayName = displayName;
    }

    public Boolean getIsActive() {
        return isActive;
    }

    public void setIsActive(Boolean isActive) {
        this.isActive = isActive;
    }

    public String getPreferences() {
        return preferences;
    }

    public void setPreferences(String preferences) {
        this.preferences = preferences;
    }

    public OffsetDateTime getCreatedAt() {
        return createdAt;
    }

    public void setCreatedAt(OffsetDateTime createdAt) {
        this.createdAt = createdAt;
    }

    public OffsetDateTime getUpdatedAt() {
        return updatedAt;
    }

    public void setUpdatedAt(OffsetDateTime updatedAt) {
        this.updatedAt = updatedAt;
    }
}

/* ==== Article.java ==== */

import jakarta.persistence.Column;
import jakarta.persistence.Entity;
import jakarta.persistence.Id;
import jakarta.persistence.JoinColumn;
import jakarta.persistence.ManyToOne;
import jakarta.persistence.Table;
import java.math.BigDecimal;
import java.time.LocalDate;
import java.time.OffsetDateTime;
import java.util.UUID;

/** 記事 */
@Entity
@Table(name = "articles")
public class Article {
    @Id
    @Column(name = "id")
    private UUID id;

    /** 執筆者 (users.id) */
    @ManyToOne
    @JoinColumn(name = "author_id", nullable = false)
    private User author;

    @Column(name = "title", nullable = false)
    private String title;

    @Column(name = "body", nullable = true)
    private String body;

    @Column(name = "view_count", nullable = false)
    private Integer viewCount;

    /** 有料記事の価格。money ではなく numeric を使う */
    @Column(name = "price", nullable = true)
    private BigDecimal price;

    @Column(name = "published_on", nullable = true)
    private LocalDate publishedOn;

    @Column(name = "created_at", nullable = false)
    private OffsetDateTime createdAt;

    @Column(name = "updated_at", nullable = false)
    private OffsetDateTime updatedAt;

    public UUID getId() {
        return id;
    }

    public void setId(UUID id) {
        this.id = id;
    }

    public User getAuthor() {
        return author;
    }

    public void setAuthor(User author) {
        this.author = author;
    }

    public String getTitle() {
        return title;
    }

    public void setTitle(String title) {
        this.title = title;
    }

    public String getBody() {
        return body;
    }

    public void setBody(String body) {
        this.body = body;
    }

    public Integer getViewCount() {
        return viewCount;
    }

    public void setViewCount(Integer viewCount) {
        this.viewCount = viewCount;
    }

    public BigDecimal getPrice() {
        return price;
    }

    public void setPrice(BigDecimal price) {
        this.price = price;
    }

    public LocalDate getPublishedOn() {
        return publishedOn;
    }

    public void setPublishedOn(LocalDate publishedOn) {
        this.publishedOn = publishedOn;
    }

    public OffsetDateTime getCreatedAt() {
        return createdAt;
    }

    public void setCreatedAt(OffsetDateTime createdAt) {
        this.createdAt = createdAt;
    }

    public OffsetDateTime getUpdatedAt() {
        return updatedAt;
    }

    public void setUpdatedAt(OffsetDateTime updatedAt) {
        this.updatedAt = updatedAt;
    }
}

/* ==== ArticleTag.java ==== */

import jakarta.persistence.Column;
import jakarta.persistence.Entity;
import jakarta.persistence.Id;
import jakarta.persistence.IdClass;
import jakarta.persistence.Table;
import java.time.OffsetDateTime;
import java.util.UUID;

/** 記事とタグの対応 */
@Entity
@Table(name = "article_tags")
@IdClass(ArticleTagId.class)
public class ArticleTag {
    @Id
    @Column(name = "article_id")
    private UUID articleId;

    @Id
    @Column(name = "tag")
    private String tag;

    @Column(name = "created_at", nullable = false)
    private OffsetDateTime createdAt;

    public UUID getArticleId() {
        return articleId;
    }

    public void setArticleId(UUID articleId) {
        this.articleId = articleId;
    }

    public String getTag() {
        return tag;
    }

    public void setTag(String tag) {
        this.tag = tag;
    }

    public OffsetDateTime getCreatedAt() {
        return createdAt;
    }

    public void setCreatedAt(OffsetDateTime createdAt) {
        this.createdAt = createdAt;
    }
}

/* ==== ArticleTagId.java ==== */

import java.io.Serializable;
import java.util.Objects;
import java.util.UUID;

/** article_tags の複合主キー（JPA は @IdClass に id クラスを要求する） */
public class ArticleTagId implements Serializable {
    private UUID articleId;
    private String tag;

    public UUID getArticleId() {
        return articleId;
    }

    public void setArticleId(UUID articleId) {
        this.articleId = articleId;
    }

    public String getTag() {
        return tag;
    }

    public void setTag(String tag) {
        this.tag = tag;
    }

    @Override
    public boolean equals(Object other) {
        if (this == other) {
            return true;
        }
        if (!(other instanceof ArticleTagId)) {
            return false;
        }
        ArticleTagId that = (ArticleTagId) other;
        return Objects.equals(articleId, that.articleId) && Objects.equals(tag, that.tag);
    }

    @Override
    public int hashCode() {
        return Objects.hash(articleId, tag);
    }
}

File: tests/golden/orm/jpa/postgresql/house-defaults.kt
/*
 * grabado が生成した Jakarta Persistence の entity（Kotlin）。
 *
 * **kotlin("plugin.jpa") が要る** —— JPA は引数の無いコンストラクタを求めるので、
 * コンストラクタ引数で書く形はプラグイン前提になる（Spring Initializr の既定）。
 * package 宣言は出さない（置き場所は生成物を受け取る側が決める）。
 */

import jakarta.persistence.Column
import jakarta.persistence.Entity
import jakarta.persistence.Id
import jakarta.persistence.IdClass
import jakarta.persistence.JoinColumn
import jakarta.persistence.ManyToOne
import jakarta.persistence.Table
import jakarta.persistence.UniqueConstraint
import java.io.Serializable
import java.math.BigDecimal
import java.time.LocalDate
import java.time.OffsetDateTime
import java.util.UUID

/** ユーザー */
@Entity
@Table(name = "users", uniqueConstraints = [UniqueConstraint(name = "users_email_key", columnNames = ["email"])])
class User(
    @Id
    @Column(name = "id")
    var id: UUID,

    /** ログイン用メールアドレス */
    @Column(name = "email", nullable = false)
    var email: String,

    @Column(name = "display_name", nullable = false)
    var displayName: String,

    @Column(name = "is_active", nullable = false)
    var isActive: Boolean,

    /** UI 設定などの任意項目 */
    @Column(name = "preferences", nullable = false)
    /* json: JPA の標準に対応する型が無いので String で出す（JSONB） */
    var preferences: String,

    @Column(name = "created_at", nullable = false)
    var createdAt: OffsetDateTime,

    @Column(name = "updated_at", nullable = false)
    var updatedAt: OffsetDateTime,
)

/** 記事 */
@Entity
@Table(name = "articles")
class Article(
    @Id
    @Column(name = "id")
    var id: UUID,

    /** 執筆者 (users.id) */
    @ManyToOne
    @JoinColumn(name = "author_id", nullable = false)
    var author: User,

    @Column(name = "title", nullable = false)
    var title: String,

    @Column(name = "body", nullable = true)
    var body: String? = null,

    @Column(name = "view_count", nullable = false)
    var viewCount: Int,

    /** 有料記事の価格。money ではなく numeric を使う */
    @Column(name = "price", nullable = true)
    var price: BigDecimal? = null,

    @Column(name = "published_on", nullable = true)
    var publishedOn: LocalDate? = null,

    @Column(name = "created_at", nullable = false)
    var createdAt: OffsetDateTime,

    @Column(name = "updated_at", nullable = false)
    var updatedAt: OffsetDateTime,
)

/** 記事とタグの対応 */
@Entity
@Table(name = "article_tags")
@IdClass(ArticleTagId::class)
class ArticleTag(
    @Id
    @Column(name = "article_id")
    var articleId: UUID,

    @Id
    @Column(name = "tag")
    var tag: String,

    @Column(name = "created_at", nullable = false)
    var createdAt: OffsetDateTime,
)

/** article_tags の複合主キー（JPA は @IdClass に id クラスを要求する） */
data class ArticleTagId(
    var articleId: UUID? = null,
    var tag: String? = null,
) : Serializable